testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeColumnMappingMode.java

```diff
@@ -290,6 +305,54 @@ public void testColumnMappingModeAddColumn(String mode)
         }
     }
 
+    @Test(groups = {DELTA_LAKE_DATABRICKS, DELTA_LAKE_OSS, DELTA_LAKE_EXCLUDE_73, DELTA_LAKE_EXCLUDE_91, PROFILE_SPECIFIC_TESTS}, dataProvider = "columnMappingDataProvider")
+    @Flaky(issue = DATABRICKS_COMMUNICATION_FAILURE_ISSUE, match = DATABRICKS_COMMUNICATION_FAILURE_MATCH)
+    public void testTrinoColumnMappingModeAddComplexColumn(String mode)
+    {
+        String tableName = "test_dl_column_mapping_mode_add_complex_column_" + randomNameSuffix();
+
+        onDelta().executeQuery("" +
+                "CREATE TABLE default." + tableName +
+                " (a_number INT)" +
+                " USING delta " +
+                " LOCATION 's3://" + bucketName + "/databricks-compatibility-test-" + tableName + "'" +
+                " TBLPROPERTIES (" +
+                " 'delta.columnMapping.mode' = '" + mode + "'" +
+                ")");
+
+        try {
+            // TODO: Insert on Trino once the connector supports id column mapping mode
+            onDelta().executeQuery("INSERT INTO default." + tableName + " VALUES 1");
+            assertThat(onDelta().executeQuery("SELECT * FROM default." + tableName)).containsOnly(row(1));
+            assertThat(onTrino().executeQuery("SELECT * FROM delta.default." + tableName)).containsOnly(row(1));
+
+            onTrino().executeQuery("ALTER TABLE delta.default." + tableName + " ADD COLUMN a_array array(integer)");
```

Comment:
Let's add a bit more nesting (array of struct, map<string, struct>, struct containing structs) to make sure that generating the new delta lake schema is accurate

Reply:
Why adding more nested fields ensures the schema is accurate?

Reply:
Fair question. This part has already been covered by other tests.
I wanted to make sure that the column mapping does indeed get added also on nested field types.

Reply:
```
spark-sql> create table tiny.test2(c1 integer) using delta location 's3://trino-ci-test/test2' tblproperties ('delta.columnMapping.mode'='id');

spark-sql> insert into tiny.test2 values 1,2;
Time taken: 1.649 seconds
spark-sql> select * from tiny.test2;
1	NULL	NULL
2	NULL	NULL

```

```
trino> alter table delta.tiny.test2 add column c2 varchar;
ADD COLUMN
trino> alter table delta.tiny.test2 add column c3 array(row(leaf1 integer, leaf2 integer));
ADD COLUMN
trino> select * from delta.tiny.test2;
 c1 |  c2  |  c3  
----+------+------
  2 | NULL | NULL 
  1 | NULL | NULL 
(2 rows)
```

```
spark-sql> insert into tiny.test2 values (3, 33, array(struct(333,3333)));
Time taken: 1.533 seconds
```

```
trino> select * from delta.tiny.test2;
 c1 |  c2  |            c3             
----+------+---------------------------
  2 | NULL | NULL                      
  1 | NULL | NULL                      
  3 | 33   | [{leaf1=333, leaf2=3333}] 
(3 rows)
```
```
spark-sql> show create table test2;
Error in query: Table or permanent view not found: test2; line 1 pos 18;
'ShowCreateTable false, [createtab_stmt#3237]
+- 'UnresolvedTableOrView [test2], SHOW CREATE TABLE, false
```



The existing code seems to have a problem with adding column mapping for arrays of structs.

Now it is 

```
"name": "c3",
            "nullable": true,
            "type": {
                "type": "array",
                "containsNull": true,
                "elementType": {
                    "type": "struct",
                    "fields": [
                        {
                            "metadata": {
                                "delta.columnMapping.id": 5,
                                "delta.columnMapping.physicalName": "col-feb99328-0cb9-4db5-8a92-40eaae0d049f"
                            },
                            "name": "leaf1",
                            "nullable": true,
                            "type": "integer"
                        },
                        {
                            "metadata": {
                                "delta.columnMapping.id": 6,
                                "delta.columnMapping.physicalName": "col-d8f32e3c-d7f2-4270-898b-4827758fa194"
                            },
                            "name": "leaf2",
                            "nullable": true,
                            "type": "integer"
                        }
                    ]
                }
            }
```

instead of 

```
{
            "name": "c3",
            "type": {
                "type": "array",
                "elementType": {
                    "type": "struct",
                    "fields": [
                        {
                            "name": "leaf1",
                            "type": "integer",
                            "nullable": true,
                            "metadata": {
                                "delta.columnMapping.id": 4,
                                "delta.columnMapping.physicalName": "col-00f1d2da-4869-4742-9b62-f1a0fff8ae39"
                            }
                        },
                        {
                            "name": "leaf2",
                            "type": "integer",
                            "nullable": true,
                            "metadata": {
                                "delta.columnMapping.id": 5,
                                "delta.columnMapping.physicalName": "col-12cb0e09-c2e7-4ee0-841e-e6270a56f06f"
                            }
                        }
                    ]
                },
                "containsNull": true
            },
            "nullable": true,
            "metadata": {
                "delta.columnMapping.id": 3,
                "delta.columnMapping.physicalName": "col-269fbe5b-e401-4d5c-8b9c-dfb903e982c0"
            }
        }
```
Notice missing 

```
"metadata": { "delta.columnMapping.id": 3, "delta.columnMapping.physicalName": "col-269fbe5b-e401-4d5c-8b9c-dfb903e982c0" }
```


Repro in spark

```
spark-sql> create table tiny.test3(c1 integer, c2 integer, c3 array<struct<leaf1 integer, leaf2 integer>>) using delta location 's3://trino-ci-test/test3' tblproperties ('delta.columnMapping.mode'='id');
```

Reply:
Good catch! I will fix and add more tests.

Reply:
I reconfirmed the generated transaction log, but it contains the expected id (3) and physicalName. I will add an integration test anyway though.

Reply:
Sorry. I'm reading again the test steps and notice that the problem was in my test scenario :( . I didn't specify the `tiny` schema when trying to see the `SHOW CREATE TABLE` statement output.

Reply:
Got a bit confused by the order of the properties - thanks for following up with https://github.com/trinodb/trino/pull/17377